staking/slash/double-sign.go

```diff
@@ -1,52 +1,301 @@
-	beneficiary common.Address,
-) Record {
-	r := Record{}
-	r.Offender = offender
-	r.Signed.Header = signedHeader
-	r.Signed.Signature = signedSignature
-	r.DoubleSigned.Header = doubleSignedHeader
-	r.DoubleSigned.Signature = doubleSignedSignature
-	r.Beneficiary = beneficiary
-	return r
-}
-
-// TODO(Edgar) Implement Verify and Apply
+	ConflictingBallots
+	// the reporter who will get rewarded
+	Evidence Evidence       `json:"evidence"`
+	Reporter common.Address `json:"reporter"`
+	Offender common.Address `json:"offender"`
+}
+
+// Application ..
+type Application struct {
+	TotalSlashed, TotalSnitchReward *big.Int
+}
+
+func (a *Application) String() string {
+	s, _ := json.Marshal(a)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (e Evidence) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Moment
+		ProposalHeader string `json:"header"`
+	}{e.Moment, e.ProposalHeader.String()})
+}
+
+// Records ..
+type Records []Record
+
+func (r Records) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (r Record) MarshalJSON() ([]byte, error) {
+	reporter, offender :=
+		common2.MustAddressToBech32(r.Reporter),
+		common2.MustAddressToBech32(r.Offender)
+	return json.Marshal(struct {
+		ConflictingBallots
+		Evidence         Evidence `json:"evidence"`
+		Beneficiary      string   `json:"beneficiary"`
+		AddressForBLSKey string   `json:"offender"`
+	}{r.ConflictingBallots, r.Evidence, reporter, offender})
+}
+
+func (e Evidence) String() string {
+	s, _ := json.Marshal(e)
+	return string(s)
+}
+
+func (r Record) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
 
 // Verify checks that the signature is valid
 func Verify(candidate *Record) error {
+
+	fmt.Println("need to verify the slash", candidate)
+
 	return nil
 }
 
-// Apply ..
-func Apply(state *state.DB, slashes []byte) error {
+var (
+	oneHundredPercent = numeric.NewDec(1)
+	fiftyPercent      = numeric.MustNewDecFromStr("0.5")
+)
+
+func delegatorSlashApply(
+	delegations staking.Delegations,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+	for _, delegator := range delegations {
+		amt := numeric.NewDecFromBigInt(
+			delegator.Amount,
+		)
+		discounted := amt.Mul(rate).TruncateInt()
+		half := numeric.NewDecFromBigInt(
+			new(big.Int).Sub(delegator.Amount, discounted),
+		).Mul(fiftyPercent)
+
+		paidOff := numeric.ZeroDec()
+
+		// NOTE Get the token trying to escape
+		for _, undelegate := range delegator.Undelegations {
+			// the epoch matters, only those undelegation
+			// such that epoch>= doubleSignEpoch should be slashable
+			if undelegate.Epoch.Cmp(doubleSignEpoch) == -1 {
+				continue
+			}
+			if paidOff.Equal(half) {
+				break
+			}
+			undelegate.Amount.SetInt64(0)
+			slashTrack.TotalSlashed.Add(slashTrack.TotalSlashed, undelegate.Amount)
+			paidOff = paidOff.Add(numeric.NewDecFromBigInt(undelegate.Amount))
```

Comment:
the undelegate.Amount is already set to 0 before.

Reply:
Also, the amount to add to paidOff shouldn't make paidOff greater than the amount to slash. Which means there should be some logic on the amount to add to paidOff, rather than simply using undelegate.Amount for the last part.